src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs

```diff
@@ -47,6 +50,50 @@ public void ReloadData()
             }
         }
 
+        static void LogChanges(Dictionary<string, HashSet<EndpointInstance>> oldInstanceMap, Dictionary<string, HashSet<EndpointInstance>> newInstanceMap)
+        {
+            var output = new StringBuilder();
+            var hasChanges = false;
+            output.AppendLine("Updating instance mapping table:");
+
+            foreach (var endpoint in newInstanceMap)
+            {
+                HashSet<EndpointInstance> existingInstances;
+                if (oldInstanceMap.TryGetValue(endpoint.Key, out existingInstances))
+                {
+                    var newInstances = endpoint.Value.Except(existingInstances).Count();
+                    var removedInstances = existingInstances.Except(endpoint.Value).Count();
+
+                    if (newInstances > 0 || removedInstances > 0)
+                    {
+                        output.AppendLine($"Updated endpoint '{endpoint.Key}': +{Instances(newInstances)}, -{Instances(removedInstances)}");
+                        hasChanges = true;
+                    }
+                }
+                else
+                {
+                    output.AppendLine($"Added endpoint '{endpoint.Key}' with {Instances(endpoint.Value.Count)}");
+                    hasChanges = true;
+                }
+            }
+
+            foreach (var removedEndpoint in oldInstanceMap.Keys.Except(newInstanceMap.Keys))
+            {
+                output.AppendLine($"Removed endpoint '{removedEndpoint}' with all instances");
```

Comment:
The wording here is a bit awkward. How about "Removed all instances of endpoint '{removedEndpoint}'" or even "Removed endpoint '{removedEndpoint}' and all instances"